### PR TITLE
Update qc logic

### LINKED ISSUE #175: When current clamp sweeps are not present, do not exit with error. Instead report a fail tag.
Instead report to lims that sweeps are not available.

### PR DESCRIPTION
**Pull request by sgratiy**
1. When current clamp sweeps are not available, do not raise exception, instead add a fail tag. This would allow qc module to complete, but will not proceed to the next module
2. If blowout sweeps are not available, do not create a fail tag because patchseq cells do not have blowout sweep.

Resolves #175

### CODE REVIEW

File: ipfx/qc_protocol.py [hunk 1]
```diff
@@ -39,9 +39,43 @@ def qc_experiment(ontology, cell_features, sweep_features, qc_criteria=None):
...
+
+    Returns
+    -------
+    num_passed_sweeps: int
+    num_sweeps: int
+
+    """
+    num_passed_sweeps = 0
+    for ss in sweep_states:
+        if ss["passed"] is True:
```
Comment (njmei) [inline ipfx/qc_protocol.py]: I make this mistake a lot too. You can just do `if ss["passed"]:`

File: ipfx/qc_feature_extractor.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+    """
+    if manual_values is None:
+        manual_values = {}
+
+    output_data = {}
+    tag_list = []
+    ontology = data_set.ontology
+    # measure blowout voltage
+    try:
```
Comment (njmei) [inline ipfx/qc_feature_extractor.py]: I wonder if it's worthwhile to separate each of these `try` and `except` blocks into their own functions? Your comments seems to suggest that there could be a `measure_blowout_voltage`, `measure_electrode_zero`, `measure_clamp_seal`, `measure_input_and_series_resistance` functions.

File: ipfx/qc_feature_extractor.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                msg = "Using manual initial access resistance"
+                logging.info(msg)
+                tag_list.append(msg)
+    #
+    output_data['input_resistance_mohm'] = ir
+    output_data["initial_access_resistance_mohm"] = sr
+
+    sr_ratio = None # input access resistance ratio
+    if ir is not None and sr is not None:
+        sr_ratio = sr / ir
```
Comment (njmei) [inline ipfx/qc_feature_extractor.py]: Do you need to check that `ir` is not zero? Or does that never happen?

File: ipfx/qc_feature_extractor.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+            if sr is not None:
+                msg = "Using manual initial access resistance"
+                logging.info(msg)
+                tag_list.append(msg)
+    #
+    output_data['input_resistance_mohm'] = ir
+    output_data["initial_access_resistance_mohm"] = sr
+
+    sr_ratio = None # input access resistance ratio
+    if ir is not None and sr is not None:
```
Comment (njmei) [inline ipfx/qc_feature_extractor.py]: I think this can just be: `if ir and sr:`

File: ipfx/sweep_props.py [hunk 5]
```diff
@@ -64,3 +64,32 @@ def extract_sweep_features_subset(feature_names, sweep_features):
...
+    -------
+    num_passed_sweeps: int
+        number of sweeps passed QC
+    num_sweeps: int
+        number of sweeps QCed
+
+    """
+    num_passed_sweeps = 0
+    for ss in sweep_states:
+        if ss["passed"] is True:
```
Comment (njmei) [inline ipfx/sweep_props.py]: `if ss.get("passed"):`

File: ipfx/qc_feature_evaluator.py [hunk 6]
```diff
@@ -183,8 +183,7 @@ def qc_current_clamp_sweep(ontology, sweep, qc_criteria=None):
 
 def evaluate_blowout(blowout_mv, blowout_mv_min, blowout_mv_max, fail_tags):
```
Comment (njmei) [inline ipfx/qc_feature_evaluator.py]: You should probably add a docstring describing this new behavior?

File: ipfx/qc_feature_evaluator.py [hunk 7]
```diff
@@ -50,9 +50,14 @@ def qc_experiment(ontology, cell_features, sweep_features, qc_criteria=None):
         logging.warning(msg)
         cell_state["fail_tags"].append(msg)
 
-    return cell_state, sweep_states
+    if cell_state["fail_tags"]:
```
Comment (njmei) [inline ipfx/qc_feature_evaluator.py]: `if cell_state.get("fail_tags"):`

File: ipfx/qc_protocol.py (hunk 1 above)
Comment (dyf) [reply ipfx/qc_protocol.py]: If "passed" isn't in the dictionary, schema validation has gone wrong.  This should not be permissive.

File: ipfx/sweep_props.py (hunk 5 above)
Comment (dyf) [reply ipfx/sweep_props.py]: this should also not be permissive, IMO

File: ipfx/qc_feature_evaluator.py (hunk 7 above)
Comment (dyf) [reply ipfx/qc_feature_evaluator.py]: same

File: ipfx/qc_protocol.py (hunk 1 above)
Comment (njmei) [reply ipfx/qc_protocol.py]: Ah, I see!

File: ipfx/qc_feature_extractor.py (hunk 2 above)
Comment (sgratiy) [reply ipfx/qc_feature_extractor.py]: I have refactored this block of code  so that each feature is evaluated by the corresponding function. Thanks.

**Pull request merged**